core/src/main/java/com/datastax/oss/driver/internal/core/util/Reflection.java

```diff
@@ -45,11 +45,10 @@
    *
    * @return null if the class does not exist.
    */
-  public static Class<?> loadClass(String className) {
+  public static Class<?> loadClass(ClassLoader classLoader, String className) {
```

Comment:
With OSGi, `Thread.currentThread().getContextClassLoader()` is able to load the class (i.e. `ExpontentialReconnectionPolicy`) just fine, but its a different class loader than what was used to initially load the super type (i.e. `ReconnectionPolicy`) we are checking against which was loaded with the bundle it belongs to (the driver core bundle).   To work around that, I changed this to pass in the `ClassLoader` of the super type (see line 180, 186), but this will not work for custom implementations that are not part of the core driver bundle.

In the general case, none of this is an issue for users unless they are running in an embedded runtime (like osgi or a web app container) or doing something tricky with classloaders.

I think if we were to provide a programmatic builder option, such as `Session.builder().withClassLoader(..)` that was only used for classes that aren't part of the core driver, that could be a solution.

Reply:
> With OSGi, Thread.currentThread().getContextClassLoader() is able to load the class (i.e. ExpontentialReconnectionPolicy) just fine, but its a different class loader than what was used to initially load the super type (i.e. ReconnectionPolicy) we are checking against which was loaded with the bundle it belongs to (the driver core bundle).

So what's happening then? The two classes are loaded, does the `isAssignableFrom` call fail?

If OSGI is too complicated (and after reading a bit on it, it seems like it is), then I agree with your suggestion of adding the "user classloader" as a builder argument. But if it's not specified, we should keep the current behavior (context class loader, then `Reflection`'s loader), because I believe it's the right thing to do in a JEE container, if the driver is in the boot classpath. I'll run a test in Tomcat to confirm this.

Reply:
exactly, `isAssignable` from fails because the classloaders used between the two classes is different.   I'll revert it back and add a classloader builder option and we can document that detail in the manual somewhere as well.

Reply:
Hrmm, thinking about this some more, I think you are right.

If you use `Thread.currentThread.getContextClassLoader()` by default, I expect it to work in the general case and also in a JEE container if the driver is in the boot classpath or loaded in the same webapp.

You could do the same thing with OSGi by loading the driver bundle in the boot classpath as well, which I think could be reasonable, but I think it also means that all libraries the driver depends on also have to be loaded in the boot classpath.  That's a little different than a webapp, where you can have a ton of jars loaded in the webapp.

I'll do a little more research on how typical this is in OSGi land, but maybe that is the right approach.

Reply:
Oh, i got something working and it didn't require something crazy like that (well at least it doesn't seem as crazy to me).  Will push up shortly.

Reply:
For now, I just changed the code to use `Class.forName` with no `ClassLoader` input provided to using the calling class' `ClassLoader`, which seems to work fine as well.  However, I think it may not work in the case where the policy comes from another bundle, which may be solved by using `DynamicImport-Package: *` (see: [this](http://billhiggins.us/blog/2011/02/24/using-reflection-in-osgi/).  Playing with that now.

Reply:
Confirmed, `Dynamic-ImportPackage: *` was needed to allow the core driver to import code from other bundles via reflection.  I added that and a test to exercise it (10be2cd).  I think that will solve our problem for OSGi, but we may have more to do with `ClassLoader`, maybe we can attack that in a future ticket, or see if it becomes a problem for users?  We may be ok doing nothing.

Reply:
> isAssignableFrom fails because the classloaders used between the two classes is different

I don't understand, this is with a built-in policy implementation right? How can two types that belong to the same JAR get loaded by different classloaders?

Regarding `Dynamic-ImportPackage: *`, as far as I understand this is the nuclear option, so I still see a value in allowing users to pass their classloader to the builder.

Reply:
> How can two types that belong to the same JAR get loaded by different classloaders?

They are loaded by different classloaders because one of them is loaded via reflection using the given `ClassLoader` and the other isn't.   

i.e.:

`ReconnectionPolicy.class` that gets passed to `Reflection.buildFromConfig` is referenced in code in `DefaultDriverContext.buildReconnectionPolicy()`.  Therefore it uses the classloader for the bundle of java-driver-core.

`ExpontentialReconnectionPolicy` gets loaded by the `ClassLoader` via reflection using the `ClassLoader` specified as an input to `Class.forName`.

* `Thread.currentThread().contextClassLoader()` is an `Launcher$AppClassLoader`, its associated with the main thread of the app, which can load classes from bundles.
* `Class.forName` without a `ClassLoader` uses the `ClassLoader` of the caller, `Reflection` (the caller) which was loaded with the bundle, so its the same `ClassLoader` as `ExponentialReconnectionPolicy`, but it does not have access to load code from outside bundles, unless using `DynamicImport-Package: *`.

One solution I just realized is that you could just load the supertype using reflection too using the same `ClassLoader` and then do the `isAssignableFrom` check.  `Thread.currentThread().contexClassLoader()` would actually work for that and may not require dynamic import.  I'll look into that.   That might not be a 100% solution (which is why we may need a way for user to provide ClassLoader), but makes the issue less pressing.

Reply:
Hah, no that doesn't work either, because the resolution of the constructor fails because the `ClassLoader` used to load the class is different than that of the arguments being checked against:

> java.lang.IllegalArgumentException: Expected class ExponentialReconnectionPolicy (specified by advanced.reconnection-policy.class) to have an accessible constructor with arguments (interface com.datastax.oss.driver.api.core.context.DriverContext)

I think the rules we need to work with are:

1. The same `ClassLoader` used to load the super type, (in this case the bundle classloader) must be used to load the input class via reflection.  (So `Class.forName(className)` makes the most sense)

2. The `ClassLoader` loading the class via reflection needs to have access to the class its trying to load.  My current way around this was to use `DynamicImport-Package: *`.

So I really think the best option is to not provide a `ClassLoader` to `Class.forName` (so we use the calling class' `ClassLoader` and use `DynamicImport-Package`.

We could try loading everything in `buildFromConfig` using Reflection (including `DriverContext` for constructor resolution), and even then it may not work and it just feels wrong.  In any case i'm trying it to see what happens.

Reply:
> We could try loading everything in buildFromConfig using Reflection (including DriverContext for constructor resolution), and even then it may not work and it just feels wrong. In any case i'm trying it to see what happens.

That doesn't work.

Reply:
I realized that Akka would likely have the same issue and came across [this](https://doc.akka.io/docs/akka/2.5/general/configuration.html#a-word-about-classloaders):

> In several places of the configuration file it is possible to specify the fully-qualified class name of something to be instantiated by Akka. This is done using Java reflection, which in turn uses a ClassLoader. Getting the right one in challenging environments like application containers or OSGi bundles is not always trivial, the current approach of Akka is that each ActorSystem implementation stores the current thread’s context class loader (if available, otherwise just its own loader as in this.getClass.getClassLoader) and uses that for all reflective accesses. This implies that putting Akka on the boot class path will yield NullPointerException from strange places: this is not supported.

I noticed that their manifest also uses `DynamicImport-Package: *` ([sbt config](https://github.com/akka/akka/blob/master/project/OSGi.scala#L36)).

Reply:
Looked more into Akka's class loading behavior today.  Here's how it resolves the `ClassLoader` by default, which matches the documentation from the previous comment:

https://github.com/akka/akka/blob/75f4e2dae297b3e04b309de8709d2dc0abf92be9/akka-actor/src/main/scala/akka/util/Reflect.scala#L162-L178

The gist is, default to `Thread.currentThread().getContextClassLoader()` and if that is null, use the ClassLoader of the caller class (with some exceptions).  In the general case, it will use `getContextClassLoader()`, but that can be overridden when creating the `ActorSystem`, which is exactly what akka-osgi does:

https://github.com/akka/akka/blob/75f4e2dae297b3e04b309de8709d2dc0abf92be9/akka-osgi/src/main/scala/akka/osgi/OsgiActorSystemFactory.scala#L53-L63

It creates a [`BundleDelegatingClassLoader`](https://github.com/akka/akka/blob/75f4e2dae297b3e04b309de8709d2dc0abf92be9/akka-osgi/src/main/scala/akka/osgi/BundleDelegatingClassLoader.scala) that falls back on `ActorSystem`s `ClassLoader`.

This is all very complex, but I think we should probably follow suit on the idea of allowing the user to pass in the `ClassLoader` as a builder option, and otherwise using `Thread.currentThread().getContextClassLoader()` by default.  I don't think we need to go as far as creating a custom OSGi module for using the driver with OSGi, but we can document how to provide a class loader that works.

Reply:
I added `SessionBuilder.withClassLoader(..)` in e0ba792.  I also added a documentation section to our OSGi docs similar to the akka documentation.  I think this is probably the best solution, the default `ClassLoader` resolution is the same as it was from the beginning, and now the user has the option to override it if they need to.  This behavioris also very close to how to akka handles things, short of providing a `driver-osgi` module, which I think might be overkill.